docs/adr/0009-async-rust.md

```diff
@@ -0,0 +1,151 @@
+What should our async strategy be?
+
+### Scope
+
+This ADR discusses what our general policy on wrapper code should be.
+It does not cover how we should plan our work.
+If we decide to embrace async Rust, we do not need to commit to any particular timeline for actually switching to it.
+
+### Desktop and gecko-js
+
+On desktop, we can't write async wrappers because it's not possible in Javascript.
+Instead we use a strategy where every function is automatically wrapped as async in the C++ layer.
+Using a config file, it's possible to opt-out of this strategy for particular functions/methods.
+
+### Android-components
+
+In Kotlin, the async wrapper layer currently lives in `android-components`.
+For the purposes of this ADR, it doesn't really matter, and this ADR will not make a distinction between wrapper code in our repo and `android-components`.
+
+## How it would work
+
+### SQLite queries
+
+One of the reasons our code currently blocks is to run SQLite queries.
+https://github.com/mozilla/uniffi-rs/pull/1837 has a system to run blocking code inside an async function.
+It would basically mean replacing code like this:
+
+```kotlin
+    override suspend fun wipeLocal() = withContext(coroutineContext) {
+        conn.getStorage().wipeLocal()
+    }
+```
+
+with code like this:
+```rust
+
+    async fn wipe_local() {
+        self.queue.execute(|| self.db.wipe_local()).await
+    }
+```
+
+We would need to merge #1837, which is currently planned for the end of 2023.
+
+### Locks
+
+Another reason our code blocks is to wait on a `Mutex` or `RWLock`.
+There are a few ways we could handle this:
+
+* The simplest is to continue using regular locks, inside a `execute()` call, which would be very similar to our current system.
+* We could also consider switching to `async_lock` and reversing the order: lock first, then make a `execute()` call.
+  This may be more efficient since the async task would suspend while waiting for the lock rather than blocking a thread
+* We could also ditch locks and use [actors and channels](https://ryhl.io/blog/actors-with-tokio/) to protect our resources.
+  It's probably not worth rewriting our current components to do this, but this approach might be useful for new components.
+
+### Network requests
+
+The last reason we block is for network requests.
+To support that we would probably need some sort of "async viaduct" that would allow consumer applications to choose either:
+- Use async functions from the `reqwest` library.
+  This matches what we currently do for `firefox-ios`.
+- Use the foreign language's network stack via an async callback interface.
+  This matches what we currently do for `firefox-android`.
+  This would require implemnenting https://github.com/mozilla/uniffi-rs/issues/1729, which is currently planed for the end of 2023.
+
+## Decision Drivers
+
+## Considered Options
+
+* **(A) Experiment with async Rust**
+
+* Pick a small component like `tabs` or `push` and use it to test our async Rust.
```

Comment:
🙋🏼‍♀️ I volunteer `remote_settings` as ~a tribute~ another test component!

* It has a small API surface.
* It could be a good "launch customer" for async networking.
* It uses a lock to keep track of some client state, and races there shouldn't be catastrophic.
* It doesn't do any persistence (e.g., to SQLite) directly.
* It has a [small Android Components wrapper](https://github.com/mozilla-mobile/firefox-android/blob/44e0f1b7ae93850a58044743be5173a8e74d9582/android-components/components/support/remotesettings/src/main/java/mozilla/components/support/remotesettings/RemoteSettingsClient.kt), that won't be too disruptive to migrate.

Reply:
First off: Thanks for offering this!  Secondly: maybe we should be thinking about this component-by-component.  

Async makes sense for `remote_settings` so that's a natural place to start exploring.

Async seems to make the least sense for our syncable-storage components.  I wonder if we should just lean into saying the non-async API is the "real" API.  We could still have an async wrapper class, but this would be more of an add-on.  In practice, this would mean that when publish documentation, we would create detailed documentation for the sync API while the async wrapper would just have some text like "all methods work the same as LoginsStore, but are wrapped with `withContext` to present an async API"